src/payment/strategies/cardinal/cardinal-client.ts

```diff
@@ -96,36 +96,34 @@ export default class CardinalClient {
         return this._getClientSDK()
             .then(client => {
                 return new Promise<ThreeDSecureToken>((resolve, reject) => {
-                    client.on(CardinalEventType.Validated, (data: CardinalValidatedData, jwt: string) => {
+                    client.on(CardinalEventType.Validated, (data: CardinalValidatedData, jwt?: string) => {
                         client.off(CardinalEventType.Validated);
-                        if (!data.ActionCode) {
-                            if (data.ErrorNumber > 0) {
-                                reject(new StandardError(data.ErrorDescription));
+                        if (!jwt) {
+                            reject(new StandardError(data.ErrorDescription ? data.ErrorDescription : 'An error was encountered while processing the transaction.'));
+                        } else {
+                            if (!data.ActionCode) {
+                                resolve({ token: jwt });
                             }
-                            resolve({ token: jwt });
-                        }
 
-                        switch (data.ActionCode) {
-                            case CardinalValidatedAction.Success:
-                                resolve({ token: jwt });
-                                break;
-                            case CardinalValidatedAction.NoAction:
-                                if (data.ErrorNumber > 0) {
-                                    reject(new StandardError(data.ErrorDescription));
-                                } else if (!data.Payment || !data.Payment.ExtendedData || data.Payment.ExtendedData.SignatureVerification !== CardinalSignatureVerification.Yes) {
-                                    reject(new StandardError('Transaction signature could not be validated.'));
-                                } else {
+                            switch (data.ActionCode) {
```

Comment:
Can you make sure that you either return early or wrap your statements in `else` so that they don't get executed if `if` condition matches? For example: 

```ts
if (!data.ActionCode) {
    resolve({ token: jwt });
}

switch (data.ActionCode) {
    case CardinalValidatedAction.Success:
        // Something

    case CardinalValidatedAction.NoAction:
        // Something
}
```

With this, the entire `switch` block gets executed even if `(!data.ActionCode)` condition matches. What you should do instead is to return early (or wrap in `else`). i.e.:

```ts
if (!jwt) {
    return reject(new StandardError(data.ErrorDescription ? data.ErrorDescription : 'An error was encountered while processing the transaction.'));
}

if (!data.ActionCode) {
    return resolve({ token: jwt });
}

switch (data.ActionCode) {
case CardinalValidatedAction.Success:
    return resolve({ token: jwt });

case CardinalValidatedAction.NoAction:
    if (data.ErrorNumber > 0) {
        return reject(new StandardError(data.ErrorDescription));
    }
    
    if (!data.Payment || !data.Payment.ExtendedData || data.Payment.ExtendedData.SignatureVerification !== CardinalSignatureVerification.Yes) {
        return reject(new StandardError('Transaction signature could not be validated.'));
    }

    return resolve({ token: jwt });

case CardinalValidatedAction.Failure:
    return reject(new StandardError('User failed authentication or an error was encountered while processing the transaction.'));


case CardinalValidatedAction.Error:
    return reject(new StandardError(data.ErrorDescription));
}
```